### PR TITLE
Regression Fixes

### LINKED ISSUE #4019: Version 6.0.0-rc.2 regressions
After installing the latest release candidate there are some regressions in term of typing results.

### System:
 - OS: macOS 12.6.5
 - CPU: (10) arm64 Apple M1 Pro
 - Memory: 168.86 MB / 32.00 GB
 - Shell: 5.8.1 - /bin/zsh
 
### Binaries:
 - Node: 18.13.0 - ~/.nvm/versions/node/v18.13.0/bin/node
 - Yarn: 1.22.19 - ~/.nvm/versions/node/v18.13.0/bin/yarn
 - npm: 9.4.2 - ~/.nvm/versions/node/v18.13.0/bin/npm

## Regressions

1. `ref` and HTMLAttributes inference is broken
     - `ref` typing is `any`
     - while using `as`, `ref` is not a valid prop
     - using `as="button"` on a `div` does not correctly infer `onClick` for instance
     - I've identified [this change](https://github.com/styled-components/styled-components/compare/v6.0.0-rc.1...v6.0.0-rc.2#diff-adc775db811d749ad367a5fd61454d85f453f549d088fd97ea1575c3eb3290f4R161-R163) by @justinbhopper as the cause. Rolling back the change solves the issue.

2. Prop inference from `css<props>` is not working on built components.

```typescript

const style = css<{prop?: boolean}>``;

const Box = styled.div`
    ${style}
`

const Usage = () => (
    <Box prop>something</Box>
)
```

In this case `prop` throws 'property does not exists on type...`.  Because the types are not generated correctly when built. Had the fix and lost it down the road 😩... 

3. The new `attrs` implementation by @justinbhopper breaks `css<Props>` with `.attrs` usage and HTML props within `.attrs` factory
    - `attrs` function props are not typed. In the below example `$textVariant` is `any`,
    ```typescript
    styled.label.attrs<TextProps>(({ $textVariant = 'footnote' }) => ({$textVariant}))
    ```
    - I've isolated the above problem to [this line](https://github.com/styled-components/styled-components/commit/d31b393c05947563e804b83b109b845d5fc00220#diff-330e887239a86c15b1f5d462583b1ac0f48a8cac1b2a12b0ea1dc015dc75757fR98) and fixed it by replacing the default typing to `((props: ExecutionContext & Props) => Props)`
    - The `css<Props>` problem, is caused by the same line and fixed by the same replacement
    - The HTML attributes within the factory are completely missing can be fixed by `Partial<OuterProps>` to the attrs params factory

Otherwise everything works as expected, can't tell if it is faster overall after the latest PRs. Prop inference with Storybook and `forwardRef` components still don't work...

### PR DESCRIPTION
**Pull request by justinbhopper**
This is a large overhaul of the `IStyledComponent` interface.  I took some techniques were taken from the DefinitelyTyped v5 type definitions as a way to solve inference and ambiguity problems that the current definitions are suffering from.

Some notable things:

1. Performance is greatly improved.  Compilation of the repo went from 8sec to 3sec.
2. ~It is not possible to disallow unknown props in `attrs()` (both the factory and non-factory overload).  The reason is because we depend on TS to infer the values they provide in order to know which attrs become optional.  I was not able to make TS both infer and enforce restrictions on the type, and any attempt to do so just caused it to resort to the `any` type on the props.~

Fixes #4019, #4022, and #4027

### CODE REVIEW

File: packages/styled-components/src/test/types.tsx [hunk 1]
```diff
@@ -266,12 +284,53 @@ const StyledDiv = styled.div``;
...
+// bar should be typed boolean
+const StyledCustomComponentWithAttrs = styled(
+  CustomComponentWithRequiredProps
+).attrs((props) => ({
+  foo: props.foo === 'black' ? 'red' : 'black',
+  hello: (world) => props.hello(world),
+  bar: !props.bar,
+}));
+
+const StyledCustomComponentWithBadAttrs = styled(
```
Comment (DonnyVerduijn) [inline packages/styled-components/src/test/types.tsx:325]: this needs a @ts-expect-error comment

Comment (justinbhopper) [reply packages/styled-components/src/test/types.tsx]: Fixed this, and also updated the negative test case examples

**Comment by DonnyVerduijn** [comment]
LGTM

**Review by DonnyVerduijn — COMMENTED** [review]
The regressions in v6.0.0-rc.1 have been resolved in this PR. From giving it a quick look everything looks okay to me. I'm not aware of any performance degradation in terms of type inference, but i didn't investigate.

**Comment by probablyup** [comment]
Looking at this locally and I see some TS errors in `types.tsx`

**Review by probablyup — APPROVED** [review]
Awesome

**Pull request merged**